doc/spec/namespace.md

```diff
@@ -0,0 +1,149 @@
+location and certificate information of a repository based on DNS, HTTP requests
+and other means.
+
+A repository should always be referenced by its fully qualified name. If a
+client presents a shortened name to a user, that name should be fully
+expanded based on rules defined by the client before contacting a remote
+server. When mirroring repositories, the original name for a repository must be
+used. Changing a repository name is equivalent to copying or moving and will not
+be referenced by the original repository.
+
+## Terminology
+
+- *Global Namespace* - The full set of referenceable names.
+- *Name* - A fully qualified string containing both the domain and resource name
+- *Repository* - A collection of objects under the same name within the
+namespace.
+- *Namespace* - also *Namespace Scope* - A collection of repositories with a
+common name prefix and set of services including registry api, index, and trust
+context.
+- *Short Name* - A name which does not contain a domain and requires
+expansion to a fully qualified name before resolving to a repository.
+
+## Format
+A name consists of two parts, a DNS host name plus a repository path. The host
+name follows the DNS rules without change. The total length of a name is 255
+characters however there is no specific limitation on DNS or path components.
+
+### Name Grammar
+~~~
+<name> ::= <hostname>"/"<path>
+<hostname> ::= <host-part>*["."<host-part>][":"<port>]
+<path> ::= <path-part>*["/"<path-part>]
+<host-part> ::= <regexp "[a-z]([-]?[a-z0-9])*">
+<port> ::= <number 1 to 65535>
+<path-part> ::= <regexp "[a-z0-9]([._-]?[a-z0-9])*">
+~~~
+
+## Discovery
+The discovery process involves resolving a name into namespace scoped metadata.
+The namespace metadata contains the full set of information needed to fetch and
+verify content associated with the namespace repository. The metadata includes
+list of registry api endpoints, the trust model, and search index. The discovery
+process should not be considered secure and therefore certificates retrieved as
+part of the discovery process should be verified before trusting.
+
+Discovery can be defined as...
+`<fully qualified name> -> scope([<registry api endpoint>, ...], <certificate>, <search index>, ...)`
+
+or in Go as...
+~~~go
+type NamespaceResolver interface {
+	Resolve(name string) Namespace
+}
+~~~
+
+### Default Method
+
+The first element of the namespace is extracted and used as the domain name for
+resolution. The domain name should be used unmodified.
+
+`<fully qualified name> -> <domain>/<path>`
+
+#### HTTPS
+The discovery related metadata will be fetched via HTTPS from the DNS resolved
+location using the remaining namespace path elements as the HTTP path in a GET
+request. A discovery request url would be in the format
+`https://<domain>/<name>?docker-discovery=1`
+
+For example, “example.com/foo/bar” would create a url
+“https://example.com/foo/bar?docker-discovery=1”
+
+##### HTML Response
```

Comment:
Yes, I think at this point I can scope it down to what is expected to be used for the initial version targeted for 1.7 and possible extended behavior in the future.